File: php/src/Hqc/GF256.php
<?php

declare(strict_types=1);

namespace PQC\Hqc;

/**
 * GF(2^8) arithmetic with irreducible polynomial x^8 + x^4 + x^3 + x^2 + 1 (0x11D).
 *
 * Generator alpha = 2 (primitive element).
 */
final class GF256
{
    private const GF_POLY = 0x11D;
    private const GF_GEN = 2;
    private const GF_MUL_ORDER = 255;

    /** @var int[] exp table (doubled for convenience) */
    private static array $exp = [];

    /** @var int[] log table */
    private static array $log = [];

    private static bool $initialized = false;

    /**
     * Initialize the exp/log lookup tables.
     */
    public static function init(): void
    {
        if (self::$initialized) {
            return;
        }

        self::$exp = array_fill(0, 512, 0);
        self::$log = array_fill(0, 256, 0);

        $x = 1;
        for ($i = 0; $i < 255; $i++) {
            self::$exp[$i] = $x;
            self::$exp[$i + 255] = $x; // wrap-around
            self::$log[$x] = $i;
            $x <<= 1;
            if ($x >= 256) {
                $x ^= self::GF_POLY;
            }
        }
        self::$log[0] = 0; // convention
        self::$exp[510] = self::$exp[0];

        self::$initialized = true;
    }

    /**
     * Addition in GF(2^8) = XOR.
     */
    public static function add(int $a, int $b): int
    {
        return $a ^ $b;
    }

    /**
     * Multiplication in GF(2^8) via log/exp tables.
     */
    public static function mul(int $a, int $b): int
    {
        if ($a === 0 || $b === 0) {
            return 0;
        }
        self::init();
        return self::$exp[self::$log[$a] + self::$log[$b]];
    }

    /**
     * Multiplicative inverse in GF(2^8).
     */
    public static function inv(int $a): int
    {
        if ($a === 0) {
            return 0;
        }
        self::init();
        return self::$exp[255 - self::$log[$a]];
    }

    /**
     * Power in GF(2^8): a^n.
     */
    public static function pow(int $a, int $n): int
    {
        if ($a === 0) {
            return ($n === 0) ? 1 : 0;
        }
        self::init();
        $logA = self::$log[$a];
        $logResult = ($logA * $n) % 255;
        if ($logResult < 0) {
            $logResult += 255;
        }
        return self::$exp[$logResult];
    }

    /**
     * Division in GF(2^8): a / b.
     */
    public static function div(int $a, int $b): int
    {
        if ($b === 0) {
            throw new \InvalidArgumentException('GF256 division by zero');
        }
        if ($a === 0) {
            return 0;
        }
        self::init();
        $logDiff = self::$log[$a] - self::$log[$b];
        if ($logDiff < 0) {
            $logDiff += 255;
        }
        return self::$exp[$logDiff];
    }
}

File: rust/hqc/src/gf256.rs
/// GF(2^8) arithmetic with irreducible polynomial x^8 + x^4 + x^3 + x^2 + 1 (0x11D).
///
/// This is the polynomial specified by the HQC specification for Reed-Solomon encoding/decoding.
/// Generator (primitive element): alpha = 2 (i.e., x).

const GF_POLY: u16 = 0x11D;
const GF_GEN: u8 = 2;

/// Precomputed log and exp tables for GF(2^8).
struct Gf256Tables {
    exp: [u8; 512],
    log: [u8; 256],
}

/// Build the log/exp tables at compile time is not feasible with const fn restrictions,
/// so we use `lazy_static`-style initialization via a function.
fn build_tables() -> Gf256Tables {
    let mut exp = [0u8; 512];
    let mut log = [0u8; 256];

    let mut x: u16 = 1;
    for i in 0..255 {
        exp[i] = x as u8;
        exp[i + 255] = x as u8;
        log[x as usize] = i as u8;
        x <<= 1;
        if x >= 256 {
            x ^= GF_POLY;
        }
    }
    log[0] = 0; // convention
    exp[510] = exp[0];

    Gf256Tables { exp, log }
}

/// Global tables, initialized once.
static GF256: std::sync::LazyLock<Gf256Tables> = std::sync::LazyLock::new(build_tables);

/// The primitive element of GF(2^8).
pub const GENERATOR: u8 = GF_GEN;

/// Addition in GF(2^8) (XOR).
#[inline]
pub fn gf256_add(a: u8, b: u8) -> u8 {
    a ^ b
}

/// Multiplication in GF(2^8) via log/exp tables.
#[inline]
pub fn gf256_mul(a: u8, b: u8) -> u8 {
    if a == 0 || b == 0 {
        return 0;
    }
    let t = &*GF256;
    t.exp[t.log[a as usize] as usize + t.log[b as usize] as usize]
}

/// Multiplicative inverse in GF(2^8). Returns 0 if a == 0.
#[inline]
pub fn gf256_inv(a: u8) -> u8 {
    if a == 0 {
        return 0;
    }
    let t = &*GF256;
    t.exp[255 - t.log[a as usize] as usize]
}

/// Exponentiation: a^n in GF(2^8).
pub fn gf256_pow(a: u8, n: i32) -> u8 {
    if a == 0 {
        return if n == 0 { 1 } else { 0 };
    }
    let t = &*GF256;
    let log_a = t.log[a as usize] as i32;
    let mut log_result = (log_a * n) % 255;
    if log_result < 0 {
        log_result += 255;
    }
    t.exp[log_result as usize]
}

/// Division: a / b in GF(2^8). Panics if b == 0.
#[inline]
pub fn gf256_div(a: u8, b: u8) -> u8 {
    assert!(b != 0, "hqc: gf256 division by zero");
    if a == 0 {
        return 0;
    }
    let t = &*GF256;
    let mut log_diff = t.log[a as usize] as i32 - t.log[b as usize] as i32;
    if log_diff < 0 {
        log_diff += 255;
    }
    t.exp[log_diff as usize]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_exp_log_roundtrip() {
        let t = &*GF256;
        for i in 1..256u16 {
            let a = i as u8;
            let log_a = t.log[a as usize];
            let exp_log_a = t.exp[log_a as usize];
            assert_eq!(exp_log_a, a, "exp(log({})) = {}, want {}", a, exp_log_a, a);
        }
    }

    #[test]
    fn test_mul_identity() {
        for i in 0..256u16 {
            let a = i as u8;
            assert_eq!(gf256_mul(1, a), a);
        }
    }

    #[test]
    fn test_mul_inverse() {
        for i in 1..256u16 {
            let a = i as u8;
            assert_eq!(gf256_mul(a, gf256_inv(a)), 1, "{} * inv({}) != 1", a, a);
        }
    }
}
